harmonica/forward/utils.py

```diff
@@ -178,3 +187,129 @@ def distance_spherical_core(
+
+    """
+    # Get coordinates of the two points
+    longitude, latitude, height = point_p[:]
+    longitude_p, latitude_p, height_p = point_q[:]
+    # Convert angles to radians
+    longitude, latitude = np.radians(longitude), np.radians(latitude)
+    longitude_p, latitude_p = np.radians(longitude_p), np.radians(latitude_p)
+    # Compute trigonometric quantities
+    cosphi = np.cos(latitude)
+    sinphi = np.sin(latitude)
+    cosphi_p = np.cos(latitude_p)
+    sinphi_p = np.sin(latitude_p)
+    coslambda = np.cos(longitude_p - longitude)
+    # Compute prime vertical radii for both points
+    prime_vertical_radius = ellipsoid._prime_vertical_radius(sinphi)
+    prime_vertical_radius_p = ellipsoid._prime_vertical_radius(sinphi_p)
+    # Compute the Euclidean distance using the close-form formula
+    return geodetic_distance_core(
+        cosphi,
+        sinphi,
+        height,
+        cosphi_p,
+        sinphi_p,
+        height_p,
+        coslambda,
+        prime_vertical_radius,
+        prime_vertical_radius_p,
+        ellipsoid.first_eccentricity ** 2,
+    )
+
+
+def geodetic_distance_core(
+    cosphi,
+    sinphi,
+    height,
+    cosphi_p,
+    sinphi_p,
+    height_p,
+    coslambda,
+    prime_vertical_radius,
+    prime_vertical_radius_p,
+    ecc_sq,
+):
+    """
+    Core computation of distance between two points in geodetic coordinates
+
+    Parameters
+    ----------
+    cosphi, sinphi : floats
+        Cosine and sine of the latitude angle for the first point
+    height : float
+        Height over ellipsoid of the first point (in meters).
+    cosphi_p, sinphi_p : floats
+        Cosine and sine of the latitude angle for the second point
+    height_p : float
+        Height over ellipsoid of the second point (in meters).
+    coslambda : float
+        Cosine of the difference between longitudes angles of both points.
+    prime_vertical_radius : float
+        Prime vertical radius for the latitude angle of the first point.
+    prime_vertical_radius_p : float
+        Prime vertical radius for the latitude angle of the second point.
+    ecc_sq : float
+        Square of ellipsoid first eccentricity.
+
+    Returns
+    -------
+    distance : float
+        Euclidean distance between both points.
+    """
+    upward_sum = prime_vertical_radius + height
+    upward_sum_p = prime_vertical_radius_p + height_p
+    term_1 = upward_sum_p ** 2 * cosphi_p ** 2
+    term_2 = upward_sum ** 2 * cosphi ** 2
+    term_3 = 2 * upward_sum * upward_sum_p * cosphi * cosphi_p * coslambda
+    term_4 = (upward_sum_p - ecc_sq * prime_vertical_radius_p) ** 2 * sinphi_p ** 2
+    term_5 = (upward_sum - ecc_sq * prime_vertical_radius) ** 2 * sinphi ** 2
+    term_6 = (
+        2
+        * (upward_sum_p - ecc_sq * prime_vertical_radius_p)
+        * (upward_sum - ecc_sq * prime_vertical_radius)
+        * sinphi
+        * sinphi_p
+    )
+    distance_sq = term_1 + term_2 - term_3 + term_4 + term_5 - term_6
```

Comment:
Any particular reason to split this up like this and not just have each component in a line for the sum?

Reply:
Nope... I split the computation in terms in order to order my head when transcribing the equations, but sure we might want to shrink them into a single line.

Reply:
(effects of late evening coding)